File: 4week/근우/1. print-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque


class Solution:
    def printTree(self, root: Optional[TreeNode]) -> List[List[str]]:
        def _findDepth(node):
            if not node:
                return 0
            return 1 + max(_findDepth(node.left), _findDepth(node.right))

        R = _findDepth(root)
        C = 2 ** (R) - 1
        res = [["" for _ in range(C)] for _ in range(R)]
        res[0][C >> 1] = str(root.val)
        que = deque([(root, 0, C >> 1)])  # cur_node, row, col
        while que:
            cur_depth_node = []
            while que:
                cur_depth_node.append(que.popleft())
            for node, r, c in cur_depth_node:
                if node.left:
                    que.append((node.left, r + 1, c - 2 ** (R - r - 2)))
                    res[r + 1][c - 2 ** (R - r - 2)] = str(node.left.val)
                if node.right:
                    que.append((node.right, r + 1, c + 2 ** (R - r - 2)))
                    res[r + 1][c + 2 ** (R - r - 2)] = str(node.right.val)
        return res

File: 4week/남규/1. Print Binary Tree.cpp
// https://leetcode.com/problems/print-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height = 0;
    
    void dfs(TreeNode* node, int depth){
        if(node->left){
            dfs(node->left, depth + 1);
        }
        
        if(node->right){
            dfs(node->right, depth + 1);
        }
        
        height = max(height, depth);
    }
    
    vector<vector<string>> printTree(TreeNode* root) {
        dfs(root, 0);
        int n = pow(2, height + 1) - 1;
        vector<vector<string>> matrix(height + 1, vector<string>(n, ""));
        
        unordered_map<TreeNode*, pair<int, int>> m;
        queue<TreeNode*> que;
        
        que.push(root);
        m[root] = {0, (n - 1) / 2};
        
        while(!que.empty()){
            TreeNode* curr = que.front();
            que.pop();
            
            int currY = m[curr].first;
            int currX = m[curr].second;
            int val = curr->val;
            bool isMinus = false;
            if(val < 0) {
                isMinus = true;
                val = -val;
            }
            
            if(val == 0){
                string cand = "0";
                matrix[currY][currX] = cand;
            } else {
                string cand = "";
                while(val){
                    string v = "";
                    v += ('0' + (val % 10));
                    cand = v + cand;
                    val /= 10;
                }
                matrix[currY][currX] = isMinus ? "-" + cand : cand;
            }
            
            if(curr->left){
                que.push(curr->left);
                m[curr->left] = {currY + 1, currX - pow(2, height - currY - 1)};
            }
            
            if(curr->right){
                que.push(curr->right);
                m[curr->right] = {currY + 1, currX + pow(2, height - currY - 1)};
            }
        }
        return matrix;
    }
};
